src/persist-txn/src/txn_read.rs

```diff
@@ -315,33 +309,78 @@ pub(crate) struct DataSubscribe<T> {
     pub(crate) remap: DataRemapEntry<T>,
 }
 
+/// An active subscription of [`DataRemapEntry`]s for a data shard.
+#[derive(Debug)]
+pub struct DataSubscription<T: Timestamp + Lattice + Codec64> {
+    /// Metadata and current [`DataRemapEntry`] for the data shard.
+    subscribe: DataSubscribe<T>,
+    /// Channel to send [`DataRemapEntry`]s.
+    tx: mpsc::UnboundedSender<DataRemapEntry<T>>,
+}
+
 impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> DataSubscribeBlocked<T> {
     /// Unblocks the snapshot portion of this subscribe. See
     /// [`DataSnapshot::unblock_read`].
     ///
-    /// Returns `None` if the upper of the data shard is the empty antichain.
+    /// Returns a [`DataSubscribe`] and a future that must be awaited to
+    /// unblock that snapshot. The returned subscribe is `None` if the upper
+    /// of the data shard is the empty antichain.
+    #[must_use]
     #[instrument(level = "debug", fields(shard = %self.0.data_id, ts = ?self.0.as_of, empty_to = ?self.0.empty_to))]
-    pub(crate) async fn unblock_subscribe<K, V, D>(
+    pub(crate) fn unblock_subscribe<'a, K, V, D>(
         self,
         data_write: WriteHandle<K, V, T, D>,
-    ) -> Option<DataSubscribe<T>>
+    ) -> (Option<DataSubscribe<T>>, BoxFuture<'a, ()>)
```

Comment:
This is clever, but feels a little clunky. Thoughts on removing DataSubscriptionBlocked and [doing something like this instead](https://github.com/MaterializeInc/materialize/compare/main...danhhz:materialize:26812_cr?expand=1)? It's less contained for sure, but not particularly more abstraction leakage than `DataSnapshot::unblock_read` already is

Reply:
I just pulled this and pushed to this PR.